src/endian/endian_stream.hpp

```diff
@@ -0,0 +1,147 @@
+
+// Copyright (c) 2016 Steinwurf ApS
+// All Rights Reserved
+//
+// Distributed under the "BSD License". See the accompanying LICENSE.rst file.
+
+#pragma once
+
+#include <stdint.h>
+#include <cassert>
+#include <algorithm>
+
+#include <storage/storage.hpp>
+#include "big_endian.hpp"
+#include "little_endian.hpp"
+
+namespace endian
+{
+    /// The idea behind the endian_stream is to provide a stream-like interface
+    /// for accessing a fixed-size buffer.
+    /// All complexity regarding endianness is encapsulated.
+    template<class EndianType>
+    class endian_stream
+    {
+    public:
+        /// Creates an endian stream on top of a pre-allocated buffer of the
+        /// specified size
+        /// @param buffer a pointer to the buffer
+        /// @param size the size of the buffer in bytes
+        endian_stream(uint8_t* buffer, uint32_t size) :
+            m_buffer(buffer), m_size(size), m_position(0)
+        {
+            assert(m_buffer != 0);
```

Comment:
to be very c++11 and modern we could use `assert(m_buffer != nullptr)` in these asserts.